# Report 
Group 5 - Product in consideration Diacetyl (mmol/gDW/h)\
Guilherme Lobo\
Karyna Lysenko\
Rodrigo Esperança

Biologia de Sistemas - Universidade do Minho\
dezembro de 2022

### Documentação (apagar depois)
https://cobrapy.readthedocs.io/en/latest/

In [1]:
import cobra
from cobra.io import read_sbml_model
from cobra.flux_analysis import flux_variability_analysis as fva
import pandas
import mewpy
from mewpy.simulation import get_simulator
from mewpy.omics import ExpressionSet
from mewpy.omics import GIMME

In [3]:
model = read_sbml_model("sth_novo.xml")
#model

#### To find out the Diacetyl metabolite id:

In [13]:
for i in model.metabolites:
    if i.name == "Diacetyl_":
        print(i.id)

diact_c
diact_e


In [5]:
diact_c=model.metabolites.get_by_id("diact_c")
#diact_c
diact_c=model.metabolites.get_by_id("diact_e")
#diact_c
DIACTt= model.reactions.get_by_id("DIACTt")
#DIACTt
casp= model.reactions.get_by_id("EX_caspep_e")
#casp
EX_diact_e= model.reactions.get_by_id("EX_diact_e")
#EX_diact_e

## Chemical defined medium (CDM)
### Function objetive: maximization of biomass production.

##### 1 a) What is the wild-type production of the compound?


In order to have a general idea of the flux of biomass_STR and of the exchange reactions we applied the summary() function and noted the biomass production is 1.042 1/h.

In [15]:
model.summary()

In [16]:
model.optimize()["EX_diact_e"]

0.0

Conclusion: The wild-type production of the Diacetyl is 0.0 mmol/gDW/h.

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [19]:
model_fva=fva(model,fraction_of_optimum=1)
model_fva.loc["EX_diact_e"]

minimum    0.000000e+00
maximum    3.700382e-13
Name: EX_diact_e, dtype: float64

Conclusion: As the minimum and maximum are very close to each other (maximum is almost 0.0), we can conclude that the solution is robust.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [11]:
model_1c = read_sbml_model("sth_novo.xml")

In [12]:
model_1c.summary(fva=0.20)

In [13]:
model_1c.optimize()["EX_diact_e"]

0.0

Conclusion: The Diacetyl production is still 0.0 mmol/gDW/h.

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.

In [3]:
from cobra.flux_analysis import (single_gene_deletion, double_gene_deletion)

In [21]:
model_knockout = read_sbml_model("sth.xml")
for i in model_knockout.genes:
    with model_knockout:
        i.knock_out()
        #tmp=model_knockout.optimize()
        #print(tmp.fluxes, tmp.status) #producao do composto + taxa de crescimento
        print(model_knockout.optimize()["EX_diact_e"])

0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)



0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)



0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)



0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0
0.0


C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)


Conclusion: The deletion of each and any gene doesn't affect the Diacetyl production, the production of it remains 0.0 mmol/gDW/h.

## Milk medium (MM)
### Function objetive: minimization of casein consumption while constraining the growth rate to 0.64 h-1

In [2]:
model_milk = read_sbml_model("sth_novo.xml")

In [3]:
model_milk.objective = 'EX_caspep_e'

model_milk.reactions.biomass_STR.lower_bound=0.64
model_milk.reactions.biomass_STR.upper_bound=0.64

#change the conditions of the solution
model_milk.reactions.EX_lac_L_e.lower_bound=30
model_milk.reactions.EX_lac_L_e.upper_bound=40.8

model_milk.reactions.EX_lcts_e.lower_bound= (-25.11)
model_milk.reactions.EX_lcts_e.upper_bound= (-22.6)

model_milk.reactions.EX_gal_e.lower_bound=0
model_milk.reactions.EX_gal_e.upper_bound=1000

model_milk.reactions.EX_glc_D_e.lower_bound=2
model_milk.reactions.EX_glc_D_e.upper_bound=2.5

model_milk.reactions.EX_caspep_e.lower_bound=-1000
model_milk.reactions.EX_caspep_e.upper_bound=1000

model_milk.reactions.EX_mpept_e.lower_bound=0
model_milk.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_milk.reactions.EX_ala_D_e.lower_bound = 0
model_milk.reactions.EX_ala_D_e.upper_bound = 1000
model_milk.reactions.EX_ala_L_e.lower_bound = 0
model_milk.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_milk.reactions.EX_arg_L_e.lower_bound = 0
model_milk.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_milk.reactions.EX_asn_L_e.lower_bound = 0
model_milk.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_milk.reactions.EX_asp_L_e.lower_bound = 0
model_milk.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_milk.reactions.EX_cys_L_e.lower_bound = 0
model_milk.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_milk.reactions.EX_gln_L_e.lower_bound = 0
model_milk.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_milk.reactions.EX_glu_L_e.lower_bound=0
model_milk.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_milk.reactions.EX_gly_e.lower_bound=0
model_milk.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_milk.reactions.EX_his_L_e.lower_bound=0
model_milk.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_milk.reactions.EX_ile_L_e.lower_bound=0
model_milk.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_milk.reactions.EX_leu_L_e.lower_bound=0
model_milk.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_milk.reactions.EX_lys_L_e.lower_bound=0
model_milk.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_milk.reactions.EX_met_L_e.lower_bound=0
model_milk.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_milk.reactions.EX_phe_L_e.lower_bound=0
model_milk.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_milk.reactions.EX_pro_L_e.lower_bound=0
model_milk.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_milk.reactions.EX_ser_D_e.lower_bound=0
model_milk.reactions.EX_ser_D_e.upper_bound=1000
model_milk.reactions.EX_ser_L_e.lower_bound=0
model_milk.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_milk.reactions.EX_thr_L_e.lower_bound=0
model_milk.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_milk.reactions.EX_trp_L_e.lower_bound=0
model_milk.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_milk.reactions.EX_tyr_L_e.lower_bound=0
model_milk.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_milk.reactions.EX_val_L_e.lower_bound=0
model_milk.reactions.EX_val_L_e.upper_bound=1000

In [7]:
# casp_m = model_milk.reactions.get_by_id("EX_caspep_e")
# casp_m

<Reaction EX_caspep_e at 0x7fa650c7b100>

#### 1 a) What is the wild-type production of the compound?

In [5]:
model_milk.summary()

In [6]:
model_milk.optimize()["EX_diact_e"]

1.346139990204063

Conclusion: The wild-type production of the Diacetyl is 1.346 mmol/gDW/h. 

#### 1 b) Access the robustness of the presented solution using the Flux Variability Analysis approach.

In [26]:
model_milk_fva=fva(model_milk,fraction_of_optimum=1)
model_milk_fva.loc["EX_diact_e"] 

minimum    0.00000
maximum    7.05441
Name: EX_diact_e, dtype: float64

Conclusion: As the minimum and maximum are very distant to each other, we can conclude that the solution is not robust.

#### 1 c) What are the maximum compound production capabilities, guaranteeing a minimum growth rate of 20% of the wild type?

In [28]:
model_milk_20 = read_sbml_model("sth_novo.xml")

In [29]:
model_milk_20.objective = 'EX_caspep_e'

model_milk_20.reactions.biomass_STR.lower_bound=0.64
model_milk_20.reactions.biomass_STR.upper_bound=0.64

#change the conditions of the solution
model_milk_20.reactions.EX_lac_L_e.lower_bound=30
model_milk_20.reactions.EX_lac_L_e.upper_bound=40.8

model_milk_20.reactions.EX_lcts_e.lower_bound= (-25.11)
model_milk_20.reactions.EX_lcts_e.upper_bound= (-22.6)

model_milk_20.reactions.EX_gal_e.lower_bound=0
model_milk_20.reactions.EX_gal_e.upper_bound=1000

model_milk_20.reactions.EX_glc_D_e.lower_bound=2
model_milk_20.reactions.EX_glc_D_e.upper_bound=2.5

model_milk_20.reactions.EX_caspep_e.lower_bound=-1000
model_milk_20.reactions.EX_caspep_e.upper_bound=1000

model_milk_20.reactions.EX_mpept_e.lower_bound=0
model_milk_20.reactions.EX_mpept_e.upper_bound=1000

#AMINOACIDS
#Alanine
model_milk_20.reactions.EX_ala_D_e.lower_bound = 0
model_milk_20.reactions.EX_ala_D_e.upper_bound = 1000
model_milk_20.reactions.EX_ala_L_e.lower_bound = 0
model_milk_20.reactions.EX_ala_L_e.upper_bound = 1000
#Arginine
model_milk_20.reactions.EX_arg_L_e.lower_bound = 0
model_milk_20.reactions.EX_arg_L_e.upper_bound = 1000
#Aspargine
model_milk_20.reactions.EX_asn_L_e.lower_bound = 0
model_milk_20.reactions.EX_asn_L_e.upper_bound = 1000
#Aspartic Acid
model_milk_20.reactions.EX_asp_L_e.lower_bound = 0
model_milk_20.reactions.EX_asp_L_e.upper_bound = 1000
#Cystein
model_milk_20.reactions.EX_cys_L_e.lower_bound = 0
model_milk_20.reactions.EX_cys_L_e.upper_bound = 1000
#Glutamine
model_milk_20.reactions.EX_gln_L_e.lower_bound = 0
model_milk_20.reactions.EX_gln_L_e.upper_bound = 1000
#Glutamic Acid
model_milk_20.reactions.EX_glu_L_e.lower_bound=0
model_milk_20.reactions.EX_glu_L_e.upper_bound=1000
#Glycine
model_milk_20.reactions.EX_gly_e.lower_bound=0
model_milk_20.reactions.EX_gly_e.upper_bound=1000
#Histidine
model_milk_20.reactions.EX_his_L_e.lower_bound=0
model_milk_20.reactions.EX_his_L_e.upper_bound=1000
#Isoleucin
model_milk_20.reactions.EX_ile_L_e.lower_bound=0
model_milk_20.reactions.EX_ile_L_e.upper_bound=1000
#Leucine
model_milk_20.reactions.EX_leu_L_e.lower_bound=0
model_milk_20.reactions.EX_leu_L_e.upper_bound=1000
#Lysine
model_milk_20.reactions.EX_lys_L_e.lower_bound=0
model_milk_20.reactions.EX_lys_L_e.upper_bound=1000
#Methionine
model_milk_20.reactions.EX_met_L_e.lower_bound=0
model_milk_20.reactions.EX_met_L_e.upper_bound=1000
#Phenylalanine
model_milk_20.reactions.EX_phe_L_e.lower_bound=0
model_milk_20.reactions.EX_phe_L_e.upper_bound=1000
#Proline
model_milk_20.reactions.EX_pro_L_e.lower_bound=0
model_milk_20.reactions.EX_pro_L_e.upper_bound=1000
#Serine
model_milk_20.reactions.EX_ser_D_e.lower_bound=0
model_milk_20.reactions.EX_ser_D_e.upper_bound=1000
model_milk_20.reactions.EX_ser_L_e.lower_bound=0
model_milk_20.reactions.EX_ser_L_e.upper_bound=1000
#Theorine
model_milk_20.reactions.EX_thr_L_e.lower_bound=0
model_milk_20.reactions.EX_thr_L_e.upper_bound=1000
#Tryptophan
model_milk_20.reactions.EX_trp_L_e.lower_bound=0
model_milk_20.reactions.EX_trp_L_e.upper_bound=1000
#Tyrosine
model_milk_20.reactions.EX_tyr_L_e.lower_bound=0
model_milk_20.reactions.EX_tyr_L_e.upper_bound=1000
#Valine
model_milk_20.reactions.EX_val_L_e.lower_bound=0
model_milk_20.reactions.EX_val_L_e.upper_bound=1000

In [30]:
model_milk_20.reactions.biomass_STR.lower_bound=0.64*0.2
model_milk_20.reactions.biomass_STR.upper_bound=0.64*0.2 

In [31]:
model_milk_20.optimize()["EX_diact_e"] 

0.0

Conclusion: The Diacetyl production is 0.0 mmol/gDW/h.

#### 1 d) Evaluate if single gene deletions enhance the production of the compound. Rank the mutants obtained according to the compound production capacity and growth performance.

In [30]:
model_milk.reactions.biomass_STR

<Reaction biomass_STR at 0x1f555de1eb0>

In [25]:
list_variants=[]
for i in model_milk.genes:
    with model_milk:
        variant=()
        objective,growth=0,0
        gene=''
        i.knock_out()
        tmp=model_milk.optimize()
        save_diact_optimize=float(tmp["EX_diact_e"])
        gene=str(i)
        objective=model_milk.slim_optimize()
        variant = (save_diact_optimize, gene, objective)
#         print(model_milk.reactions.biomass_STR)
        list_variants.append(variant)

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning

C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning)
C:\Users\Karyna\anaconda3\lib\site-packages\cobra\util\solver.py:554: UserWarning: Solver status is 'infeasible'.
  warn(f"Solver status is '{status}'.", UserWarning

In [28]:
sorted(list_variants, reverse=True)

[(6.9877614079999395, 'CH8_0368', nan),
 (6.9877614079999395, 'CH8_0352', nan),
 (6.935134208000038, 'CH8_1177', nan),
 (6.935134208000035, 'CH8_1893', nan),
 (6.935134207999991, 'CH8_1718', nan),
 (6.935134207999991, 'CH8_1717', nan),
 (6.935134207999991, 'CH8_1716', nan),
 (6.935134207999991, 'CH8_1680', -0.008768319999999996),
 (6.93513420799999, 'CH8_1681', -0.008768319999999996),
 (6.770437551999983, 'CH8_0428', nan),
 (6.770437551999983, 'CH8_0423', nan),
 (6.48722459036736, 'CH8_0332', -0.007148042448979599),
 (6.487224590367349, 'CH8_0336', -0.007148042448979599),
 (6.487224590367349, 'CH8_0335', -0.007148042448979599),
 (6.487224590367349, 'CH8_0334', -0.007148042448979599),
 (6.487224590367349, 'CH8_0333', -0.007148042448979599),
 (6.487224590367349, 'CH8_0331', -0.007148042448979599),
 (6.487224590367349, 'CH8_0330', -0.007148042448979599),
 (6.487224590367348, 'CH8_0329', -0.007148042448979598),
 (5.300144406468101, 'CH8_0509', nan),
 (5.300144406468101, 'CH8_0503', nan),
 

Conclusion: The last output is a list of tuples where 1st value is the Diacetyl production, the 2nd is the gene that was deleted and the 3rd is the grouth performance. The list was sorted by the 1st value. We can note several things:
-the deletion of some genes turns the system infeasable -> "nan";

-the deletion of some genes, such as CH8_1680, CH8_1681, CH8_0428, etc., increases the production of Diacetyl but in terms of maximization of the objective function it gets worse, as the 3rd value gets higher in module, so the consumption of EX_caspep_e is higher in those cases;

-there are some other cases of gene deleltion, such as CH8_0316, CH8_1892, etc., qhere the flux of the objective function is the same before any deletions (-0.0030233600000000157 mmol/gDW/h) but the production of the Diacetyl is higher, around 3.18 mmol/gDW/h.

-the


#### 2) The authors of the model obtained gene expression data (RNA-seq) for this strain in CDM and milk medium (“gene_expression.xlsx”) in anaerobic conditions. For this exercise, consider the maximization of the growth rate as the objective for both media. For milk medium, set the uptake of casein as 0.004 mmol/gDW/h.

#### a) Integrate the expression data for both conditions using the GIMME algorithm. Compare the production of your compound in both conditions, as well as the flux of its production pathway(s).

In [ ]:
#alterar a função objetivo e "uptake da caseina -> "-0.004" (ver email emanuel) 
#model_milk.reactions.EX_caspep_e.lower_bound=0.004
#model_milk.reactions.EX_caspep_e.upper_bound=0.004

In [8]:
gene_expression = pandas.read_excel("gene_expression.xlsx")
gene_expression

ID         CDM        MILK
0     CH8_0001  198.412656  217.468245
1     CH8_0002  323.207616  319.916967
2     CH8_0003   53.441435   58.719433
3     CH8_0004  333.943347  349.360850
4     CH8_0005   40.002946   56.077228
...        ...         ...         ...
2156  CH8_2157    0.000000    1.674210
2157  CH8_2158  142.603093  163.920837
2158  CH8_2159  159.853040  169.367615
2159  CH8_2160  643.218883  687.317443
2160  CH8_2161   52.616722   80.981196

[2161 rows x 3 columns]

## GIMME- Chemical defined medium (CDM)

In [11]:
envcond = {
"EX_o2_e":(0,0) 
 }
#Usa-se este comando para insturir o meio anaerobio e não se coloca mais nada porque o meio ja esta defenido no modelo

simul = get_simulator(model, envcond = envcond)

#get genes and expression data
genes = gene_expression.ID.values  
conditions = ["CDM"]
exp = gene_expression[conditions].to_numpy()

# load expression module
expr = ExpressionSet(genes, conditions, exp)

# Apply parsimonious GIMME in the model
g = GIMME(simul, expr, parsimonious=True)

print(g)
print()
print(model.optimize()["EX_diact_e"])

AttributeError: optimality is not an available tolerance parameter with this solver

## GIMME- Milk medium (MM)

In [ ]:
envcond = {
"EX_caspep_e":(0.004,0.004),
"EX_o2_e":(0,0) #oxygen exchange
 }

simul = get_simulator(model_milk, envcond = envcond)

#get genes and expression data
genes = gene_expression.ID.values  
conditions = ["MILK"]
exp = gene_expression[conditions].to_numpy()

# load expression module
expr = ExpressionSet(genes, conditions, exp)

# Apply parsimonious GIMME in the model
g = GIMME(simul, expr, parsimonious=True)
print(g)
print()
print(model_milk.optimize()["EX_diact_e"])

#### b) Compare the output of the algorithm with the results obtained in exercise 1 a).